[AAQ-125] add deployment code

suzinyou opened the pull request:
Reviewer: @amiraliemami 

Estimate: 2hrs

----

# [AAQ-125] add deployment code

## Ticket

Fixes: https://idinsight.atlassian.net/browse/AAQ-125

## Description, Motivation and Context

### Goal
To allow deployment on remote test server

### Changes and why
Now we have 3 docker-compose scripts. 
1. `docker-compose.yml`: common configurations for any environment
2. `docker-compose.dev.yml`: configs for local development 
3. `docker-compose.test.yml`: configs for remote test server
This is because the test server has to connect to a remote test DB instance rather than a DB docker container.

Another change is github actions workflow `.github/workflows/docker-build-push.yml` which builds and pushes the docker images of `core_backend` and `admin_app` to our docker repository (which are private for the time being).

## How Has This Been Tested?
### 
### Github Actions workflow
Since this new action is not merged to main, I had to add this feature branch to the workflow trigger condition in order to trigger the workflow. (Adding `workflow_dispatch` option didn't show the manual run button on Github :/)

```
on:
  workflow_dispatch:
  push:
    branches:
      - main
      - AAQ-125-autodeploy-dev  # temporary code for testing 
```

- [ ] You can use [the latest run](https://github.com/IDinsight/aaq-core/actions/workflows/docker-build-push.yaml) of the `docker-build-push.yaml` workflow to test it yourself.
- [ ] Check that the newly built images have been pushed correctly to our private docker repository. 
- [ ] SSH to remote and run the `scripts/redeploy.sh` script to re-deploy the new containers


## To-do before merge

- [ ] remove `AAQ-125-autodeploy-dev` from workflow trigger condition

### What's left
I have not figured out a completely github-secret-less deployment workflow so once the images are created and pushed to docker repo, the redeployment of the new images is still **manual**.

## Checklist:

This checklist is a useful reminder of small things that can easily be forgotten.
Put an `x` in all the items that apply and remove any items that are not relevant to this PR.

- [ ] My code follows the style guidelines of this project
- [ ] I have reviewed my own code to ensure good quality
- [ ] I have tested the functionality of my code to ensure it works as intended
- [ ] I have resolved merge conflicts
- [ ] I have updated the automated tests (if applicable)
- [ ] I have written [good commit messages][1]
- [ ] I have updated the README file (if applicable)
- [ ] I have updated affected documentation (if applicable)

[1]: http://chris.beams.io/posts/git-commit/

suzinyou commented on .github/workflows/deploy.yaml:
```diff
@@ -0,0 +1,45 @@
+          webhook_type: json
+          webhook_url: ${{ env.WEBHOOK_TEST_REDEPLOY_URL }}
+          webhook_secret: ${{ env.WEBHOOK_TEST_REDEPLOY_SECRET }}
+          data: '{ "branch": "${{ github.ref_name }}" }'
```
This allows us to deploy the branch where this workflow is run.
Temporary solution to deploying test branches! (temporary because we will only have one ec2 instance to do this :/)

suzinyou commented on core_backend/app/__init__.py:
```diff
@@ -21,6 +21,8 @@ def create_app() -> FastAPI:
         "https://localhost:3000",
         "http://localhost",
         "http://localhost:3000",
+        f"http://{DOMAIN}",
+        f"http://{DOMAIN}:3000",
```
This is a workaround for the following error:
![image](https://github.com/IDinsight/aaq-core/assets/7042047/d64a89f1-34f4-4a6b-a4ad-af85fc07742c)

ziakhan04 commented:
@suzinyou, Hello! Approving the PR! The autodeploy works well :)

I have added a few comments. You can decide to address now or raise tickets.

Beyond the comments, one thing that I think is confusing right now is the dev docker compose for the following reasons -
- There is no 1 clear reason to use it. If I want to test my app, I can just run the normal docker compose with the RDS instance endpoint's URL. If I want to test and develop quickly, I will just run the apps in the terminal (python main.py / npm run dev) because of faster app reloading.
- It is also confusing to run the docker compose quickly now because of HTTPS + modifying env vars everytime I want to run locally on a postgres or an RDS.

I think we can discuss this in a tech session. Potential options -
- Not have dev docker compose at all. We can only have docker compose for prod situations. We could just run apps in the terminal for dev
- Have dedicated docker compose for dev and docker compose for prod setups. Dev could have simple hardcoded env configs for the apps and ngins (no HTTPS)

suzinyou commented on core_backend/app/__init__.py:
```diff
@@ -21,6 +21,8 @@ def create_app() -> FastAPI:
         "https://localhost:3000",
```
Good point!

suzinyou commented:
Hello Zia, thank you for this suggestion. I've linked this to the ticket you created.

I've addressed your comments also made some changes, namely 
1. In `init-letsencrypt.sh` the point of all the docker commands is to save proper SSL certificates in the volume that certbot and nginx containers will share. I removed the `--name aaq-stack` because it was causing "existing container" issues and isn't really necessary. (cc: @sidravi1 ) 
2. I checked that it works, and finally issued a certificate for our dev domain.

> @suzinyou, Hello! Approving the PR! The autodeploy works well :)
> 
> I have added a few comments. You can decide to address now or raise tickets.
> 
> Beyond the comments, one thing that I think is confusing right now is the dev docker compose for the following reasons -
> 
> * There is no 1 clear reason to use it. If I want to test my app, I can just run the normal docker compose with the RDS instance endpoint's URL. If I want to test and develop quickly, I will just run the apps in the terminal (python main.py / npm run dev) because of faster app reloading.
> * It is also confusing to run the docker compose quickly now because of HTTPS + modifying env vars everytime I want to run locally on a postgres or an RDS.
> 
> I think we can discuss this in a tech session. Potential options -
> 
> * Not have dev docker compose at all. We can only have docker compose for prod situations. We could just run apps in the terminal for dev
> * Have dedicated docker compose for dev and docker compose for prod setups. Dev could have simple hardcoded env configs for the apps and ngins (no HTTPS)
